core/src/main/scala/shapeless/annotation.scala

```diff
@@ -290,10 +384,71 @@ class AnnotationMacros(val c: whitebox.Context) extends CaseClassMacros {
     else q"_root_.shapeless.Annotations.mkAnnotations[$annTpe, $tpe, $outTpe]($outTree)"
   }
 
+  def materializeAllAnnotations[T: WeakTypeTag, Out: WeakTypeTag](typeAnnotation: Boolean): Tree = {
+    val tpe = weakTypeOf[T]
+    val annTreeOpts = getAnnotationTreeOptions(tpe, typeAnnotation)
+    
+    val wrapTpeTrees = annTreeOpts.map {
+      case Nil =>
+        mkHListTpe(Nil) -> q"(_root_.shapeless.HNil)"
+      case list =>
+        mkHListTpe(list.map(_._1)) -> list.foldRight(q"_root_.shapeless.HNil": Tree) {
+          case ((_, bound), acc) => pq"_root_.shapeless.::($bound, $acc)"
+        }
+    }
+
+    val outTpe = mkHListTpe(wrapTpeTrees.map { case (aTpe, _) => aTpe })
+    val outTree = wrapTpeTrees.foldRight(q"_root_.shapeless.HNil": Tree) {
+      case ((_, bound), acc) =>
+        pq"_root_.shapeless.::($bound, $acc)"
+    }
+
+    if (typeAnnotation) q"_root_.shapeless.AllTypeAnnotations.mkAnnotations[$tpe, $outTpe]($outTree)"
+    else q"_root_.shapeless.AllAnnotations.mkAnnotations[$tpe, $outTpe]($outTree)"
+  }
+
+  def getAnnotationTreeOptions(tpe: Type, typeAnnotation: Boolean): List[List[(Type, Tree)]] = {
+    if (isProduct(tpe)) {
+      val constructorSyms = tpe
+        .member(termNames.CONSTRUCTOR)
+        .asMethod
+        .paramLists
+        .flatten
+        .map(sym => nameAsString(sym.name) -> sym)
+        .toMap
+
+      fieldsOf(tpe).map {
+        case (name, _) =>
+          extract(typeAnnotation, constructorSyms(nameAsString(name))).collect {
+            case ann if isProduct(ann.tree.tpe) =>
+              val construct1 = construct(ann.tree.tpe)
+              (ann.tree.tpe, construct1(ann.tree.children.tail))
+          }
+      }
+    } else if (isCoproduct(tpe)) {
+      ctorsOf(tpe).map { cTpe =>
+        extract(typeAnnotation, cTpe.typeSymbol).collect {
+          case ann if isProduct(ann.tree.tpe) =>
+            val construct1 = construct(ann.tree.tpe)
+            (ann.tree.tpe, construct1(ann.tree.children.tail))
+        }
+      }
+    } else {
+      abort(s"$tpe is not case class like or the root of a sealed family of types")
+    }
+  }
+  
   def extract(tpe: Boolean, s: Symbol): List[c.universe.Annotation] = {
     if (tpe) {
       s.typeSignature match {
-        case a: AnnotatedType => a.annotations
+        case a: AnnotatedType => a.annotations.reverse
+        case c: ClassInfoType =>
+          val parents = c.parents         
+          parents.flatMap {
+            case a: AnnotatedType => a.annotations.reverse
+            case _ => Nil
+          }
+        case t: TypeRef => extract(tpe, t.sym)
```

Comment:
We should only follow aliases. And I suggest a bit of cleanup with a helper:

```scala
  def extract(tpe: Boolean, s: Symbol): List[c.universe.Annotation] = {
    def fromType(t: Type): List[c.universe.Annotation] = t match {
      case AnnotatedType(annotations, _) => annotations.reverse
      case ClassInfoType(parents, _, _) => parents.flatMap(fromType)
      case TypeRef(_, sym, _) if sym.asType.isAliasType => extract(tpe, sym)
      case _ => Nil
    }

    if (tpe) fromType(s.typeSignature)
    else s.annotations
  }
```

Reply:
sure. done.